homeassistant/components/kodi/media_player.py

```diff
@@ -327,8 +328,13 @@ def async_on_volume_changed(self, sender, data):
     @callback
     def async_on_quit(self, sender, data):
         """Reset the player state on quit action."""
+        self.hass.async_create_task(self._clear_connection())
+
+    async def _clear_connection(self, close=True):
```

Comment:
Make it a `@callback` and move the closing of the connection to be inside the exception handler in `async_ws_connect`, as that's the only place it's used. Then we don't need to create task and call it directly.

Reply:
It's the other way around - `async_ws_connect` is the only place that's *not* closing. Both `async_on_quit` and _`ping` eventually close the connection. The first is a @callback, the other async.

Reply:
Looks like I might be able to make `async_on_quit` a coroutine:
https://github.com/emlove/jsonrpc-base/blob/master/jsonrpc_base/jsonrpc.py#L90

Reply:
Even better 👍

Reply:
@emlove it looks like the version of jsonrpc-websocket that supports coroutine request handlers was never published on PyPI. Do you mind publishing it?

Reply:
(sorry, I mixed up the jsonrpc-async and jsonrpc-websocket versions. Please ignore)

Reply:
Hmm, I do need your help though. Using the latest jsonrpc-websocket, with a coroutine for `OnQuit`, I get:
```
Traceback (most recent call last):
  File "../core/venv/lib/python3.7/site-packages/jsonrpc_websocket/jsonrpc.py", line 99, in _ws_loop
    response = await self.async_receive_request(request)
  File "../core/venv/lib/python3.7/site-packages/jsonrpc_websocket/jsonrpc.py", line 38, in send_message
    await self._client.send_str(message.serialize())
  File "../core/venv/lib/python3.7/site-packages/jsonrpc_base/jsonrpc.py", line 168, in serialize
    return json.dumps(data)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type Request is not JSON serializable
```

Reply:
OK, managed to solve the dependency situation between `jsonrpc-async` and `jsonrpc-websocket` (we're using both, depending on the type of communication we have with Kodi).
Now I'm getting:
```
Traceback (most recent call last):
  File "../core/venv/lib/python3.7/site-packages/jsonrpc_websocket/jsonrpc.py", line 67, in _ws_loop
    msg = await self._client.receive()
AttributeError: 'NoneType' object has no attribute 'receive'
```
Seems like closing the connection sets `_client` to `None`, but doesn't stop the loop:
https://github.com/emlove/jsonrpc-websocket/blob/master/jsonrpc_websocket/jsonrpc.py#L116

Sorry for driving you crazy here :)

Reply:
Whoops! Looks like you're right. I've pushed up version 1.2.1 of `jsonrpc-websocket` which should hopefully fix this.

Reply:
Thanks @emlove!!!
@balloob `async_on_quit` is now a coroutine, and no task is created.
I think it's advisable to create a new beta version after this is merged, since this required updating major versions of the 3 main dependencies. It's working for me, but there are lots of Kodi setups out there.